be/src/exprs/agg/array_agg.h

```diff
@@ -16,78 +16,199 @@
+        auto& input_columns = down_cast<const StructColumn*>(ColumnHelper::get_data_column(column))->fields();
+        for (auto i = 0; i < input_columns.size(); ++i) {
+            auto array_column = down_cast<const ArrayColumn*>(ColumnHelper::get_data_column(input_columns[i].get()));
+            auto& offsets = array_column->offsets().get_data();
+            this->data(state).update(ctx, array_column->elements(), i, offsets[row_num],
+                                     offsets[row_num + 1] - offsets[row_num]);
+        }
     }
 
+    // serialize each state->column to a [nullable] array in a [nullable] struct
     void serialize_to_column(FunctionContext* ctx, ConstAggDataPtr __restrict state, Column* to) const override {
         auto& state_impl = this->data(state);
-        auto* column = down_cast<ArrayColumn*>(to);
-        column->append_array_element(state_impl.data_column, state_impl.null_count);
+        auto& columns = down_cast<StructColumn*>(ColumnHelper::get_data_column(to))->fields_column();
+        if (to->is_nullable()) {
+            down_cast<NullableColumn*>(to)->null_column_data().emplace_back(0);
+        }
+        for (auto i = 0; i < columns.size(); ++i) {
+            auto elem_size = (*state_impl.data_columns)[i]->size();
+            auto array_col = down_cast<ArrayColumn*>(ColumnHelper::get_data_column(columns[i].get()));
+            if (columns[i]->is_nullable()) {
+                down_cast<NullableColumn*>(columns[i].get())->null_column_data().emplace_back(0);
+            }
+            if ((*state_impl.data_columns)[i]->only_null()) {
+                array_col->elements_column()->append_nulls(elem_size);
+            } else {
+                array_col->elements_column()->append(
+                        *ColumnHelper::unpack_and_duplicate_const_column(elem_size, (*state_impl.data_columns)[i]), 0,
+                        elem_size);
+            }
+            auto& offsets = array_col->offsets_column()->get_data();
+            offsets.push_back(offsets.back() + elem_size);
+        }
     }
 
+    // finalize each state->column to a [nullable] array
     void finalize_to_column(FunctionContext* ctx, ConstAggDataPtr __restrict state, Column* to) const override {
-        return serialize_to_column(ctx, state, to);
+        DCHECK(to->is_nullable() || to->is_array());
+        auto& state_impl = this->data(state);
+        auto elem_size = (*state_impl.data_columns)[0]->size();
+        auto res = (*state_impl.data_columns)[0];
+        auto tmp = (*state_impl.data_columns)[0]->clone_empty();
+        if (state_impl.data_columns->size() > 1) {
+            Permutation perm;
+            Columns sort_by_columns;
+            SortDescs sort_desc(ctx->get_is_asc_order(), ctx->get_nulls_first());
+            sort_by_columns.assign(state_impl.data_columns->begin() + 1, state_impl.data_columns->end());
+            Status st = sort_and_tie_columns(false, sort_by_columns, sort_desc, &perm);
```

Comment:
just doubt: if the input's column is sorted before, how we avoid sorting again?

Reply:
generally no need to avoid sorting again, as the grouping items are not very large in generally. we can optimize it when it becomes a bottleneck.